File: source/com/barrybecker4/optimization/strategy/GeneticSearchStrategy.java
// Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.strategy;

import com.barrybecker4.common.concurrency.ThreadUtil;
import com.barrybecker4.common.format.FormatUtil;
import com.barrybecker4.common.math.MathUtil;
import com.barrybecker4.optimization.optimizee1.Optimizee;
import com.barrybecker4.optimization.parameter1.ParameterArray;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Genetic Algorithm (evolutionary) optimization strategy.
 * Many different strategies are possible to alter the population for each successive iteration.
 * The 2 primary ones that I use here are unary mutation and cross-over.
 * See Chapter 6 in "How to Solve it: Modern Heuristics" for more info.
 *
 * @author Barry Becker
 */
public class GeneticSearchStrategy extends OptimizationStrategy {

    // Percent amount to decimate the parent population by on each iteration
    private static final double CULL_FACTOR = 0.8;
    private static final double NBR_RADIUS = 0.08;
    private static final double NBR_RADIUS_SHRINK_FACTOR = 0.7;
    private static final double NBR_RADIUS_EXPAND_FACTOR = 1.1;
    private static final double NBR_RADIUS_SOFTENER = 10.0;
    private static final double INITIAL_RADIUS = 1.0;
    private static final int MAX_NBRS_TO_EXPLORE = 8;

    /** this prevents us from running forever.  */
    private static final int MAX_ITERATIONS = 100;

    /** stop when the avg population score does not improve by better than this  */
    private static final double DEFAULT_IMPROVEMENT_EPS = 0.000000000001;

    /** radius to look for neighbors in  */
    private double nbrRadius = NBR_RADIUS;

    /**
     * this is the desired number of members to be maintained in the population at any time.
     * Might not always get this many if there are duplicates or the search space is small.
     */
    private int desiredPopulationSize;

    /* If crossover breeding of genetic material is used. */
    //private boolean useCrossOver_ = false;

    /** if we don't improve by at least this amount between iterations, terminate.  */
    protected double improvementEpsilon = DEFAULT_IMPROVEMENT_EPS;


    /**
     * Constructor
     * use a hardcoded static data interface to initialize.
     * so it can be easily run in an applet without using resources.
     * @param optimizee the thing to be optimized.
     */
    public GeneticSearchStrategy( Optimizee optimizee ) {
        super(optimizee);
    }

    /**
     * //@param useCrossOver if true then create new population members using genetic crossover between parents.
     *
    public void setUseCrossOver(boolean useCrossOver) {
         useCrossOver_ = useCrossOver;
    }   */

    public void setImprovementEpsilon(double eps) {
        improvementEpsilon = eps;
    }

    /**
     * finds a local minima using a genetic algorithm (evolutionary) search.
     * Stop iterating as soon as the average evaluation score of the population
     * does not significantly improve.
     *
     * @param params the initial value for the parameters to optimize.
     * @param fitnessRange the approximate absolute value of the fitnessRange.
     * @return the optimized params.
     */
    @Override
    public ParameterArray doOptimization( ParameterArray params, double fitnessRange) {

        ParameterArray lastBest;
        desiredPopulationSize = params.getSamplePopulationSize();

        // create an initial population based on params and POPULATION_SIZE-1 other random candidate solutions.
        List<ParameterArray> population = new LinkedList<>();
        population.add(params);

        int i = 0;
        int max = 100 * desiredPopulationSize;
        while (population.size() < desiredPopulationSize && i < max) {
             ParameterArray nbr = params.getRandomNeighbor(INITIAL_RADIUS);
             if (!population.contains(nbr)) {
                 population.add(nbr);
             }
             i++;
        }
        if (population.size() <= 1)
            throw new IllegalStateException("No random neighbors found");
        //System.out.println("The population of this generation is " + population.size()
        //    + " (desired was " + desiredPopulationSize + ")");

        // EVALUATE POPULATION
        lastBest = evaluatePopulation(population, params);

        return findNewBest(params, lastBest, population);
     }

    /**
     * Find the new best candidate.
     * @return the new best candidate.
     */
    private ParameterArray findNewBest(ParameterArray params, ParameterArray lastBest,
                                       List<ParameterArray> population) {
        ParameterArray currentBest;
        int ct = 0;
        double deltaFitness;
        ParameterArray recentBest = lastBest;
        //System.out.println("findNewBest: recent best =" + recentBest);

        // each iteration represents a new generation of the population.
        do {
            int keepSize = cullPopulation(population);
            replaceCulledWithKeeperVariants(population, keepSize);

            // EVALUATE POPULATION
            currentBest = evaluatePopulation(population, recentBest);
            System.out.println("currBest = " + currentBest + " \nrecBest = " + recentBest + "        ct="+ct);

            deltaFitness = computeFitnessDelta(params, recentBest, currentBest, ct);
            System.out.println("delta fitness =" + deltaFitness + "      rbrRadius = " + nbrRadius + "  improvementEpsilon = " + improvementEpsilon);
            double factor = deltaFitness < (-1000000 * improvementEpsilon) ? NBR_RADIUS_EXPAND_FACTOR : NBR_RADIUS_SHRINK_FACTOR;
            nbrRadius *= factor;
            recentBest = currentBest.copy();

            notifyOfChange(currentBest);
            ct++;

        } while ( (deltaFitness < -improvementEpsilon)
                && !isOptimalFitnessReached(currentBest)
                && (ct < MAX_ITERATIONS));

        if (isOptimalFitnessReached(currentBest)) {
            System.out.println("stopped because we found the optimal fitness.");
        }
        else if (deltaFitness >= -improvementEpsilon) {
            System.out.println("stopped because we made no IMPROVEMENT. The delta, "
                    + deltaFitness + " was >= " + -improvementEpsilon );
        }
        else {
            System.out.println("Stopped because we exceeded the MAX ITERATIONS: " + ct);
        }
        System.out.println("----------------------- done -------------------");
        log(ct, currentBest.getFitness(), 0, 0, currentBest, FormatUtil.formatNumber(ct));
        return currentBest;
    }

    /**
     * Computes the fitness delta, but also logs and asserts that it is not 0.
     * @return the different in fitness between current best and last best.
     */
    private double computeFitnessDelta(ParameterArray params, ParameterArray lastBest,
                                       ParameterArray currentBest, int ct) {
        double deltaFitness;
        deltaFitness = (currentBest.getFitness() - lastBest.getFitness());
        assert (deltaFitness <= 0) :
                "We must never get worse in a new generation. Old fitness="
                        + lastBest.getFitness() + " New Fitness = " + currentBest.getFitness() + ".";

        //System.out.println(" ct="+ct+"  nbrRadius = " + nbrRadius + "  population size =" + desiredPopulationSize
        //                   +"  deltaFitness = " + deltaFitness+"  currentBest = " + currentBest.getFitness()
        //                   +"  lastBest = " + lastBest.getFitness());
        log(ct, currentBest.getFitness(), nbrRadius, deltaFitness, params, "---");
        return deltaFitness;
    }

    /**
     * Remove all but the best candidates. Better candidates have lower values.
     * @param population the whole population. It will be reduced in size.
     * @return the number of members that were retained.
     */
    private int cullPopulation(List<ParameterArray> population) {

        // sort the population according to the fitness of members.
        Collections.sort(population);

        // throw out the bottom CULL_FACTOR*desiredPopulationSize members - keeping the cream of the crop.
        // then replace those culled with unary variations of those (now parents) that remain.
        // @@ add option to do cross-over variations too.
        int keepSize = Math.max(1, (int)(population.size() * (1.0 - CULL_FACTOR)));

        int size = population.size();
        for (int j = size-1; j >= keepSize; j--) {
            population.remove( j );
        }
        //System.out.println("pop after culling: first = "  +  population.get(0) + " last("+population.size()+")");
        //printPopulation(population, 5);
        return keepSize;
    }

    /**
     * Replace the members of the population that were removed with variations of the ones that we kept.
     * @param population population
     * @param keepSize the number that were kept
     */
    private void replaceCulledWithKeeperVariants(List<ParameterArray> population, int keepSize) {

        int k = keepSize;
        //System.out.println("keepSize = " + keepSize + " grow current popSize of "
        //    + population.size() + " to " + desiredPopulationSize);
        while ( k < desiredPopulationSize) {

            // loop over the keepers until all replacements found. Select randomly, but skewed toward the better ones
            double rnd =  MathUtil.RANDOM().nextDouble();
            int keeperIndex = (int) (rnd * rnd * keepSize);  //k % keepSize;

            ParameterArray p = population.get(keeperIndex);

            // Add a permutation of one of the keepers.
            // Multiply the radius by m because we want the worse ones to have higher variability.
            double r = (keeperIndex + NBR_RADIUS_SOFTENER) / NBR_RADIUS_SOFTENER * nbrRadius;
            ParameterArray nbr = getNeighbor(p, r); //p.getRandomNeighbor(r);
            if (!population.contains(nbr)) {
                population.add(nbr);
                notifyOfChange(p);
            }
            k++;
        }
        //printPopulation(population, 20);
    }

    /**
     *
     * @param p parameter array to get neighbor for
     * @param rad larger radius means more distanc neighbor
     * @return a neighbor of p. If using absolute fitness, try to find a neighbor that has better fitnesss.
     */
    private ParameterArray getNeighbor(ParameterArray p, double rad) {
        ParameterArray nbr = p.getRandomNeighbor(rad);
        if (!optimizee.evaluateByComparison()) {
            // try to find a nbr with fitness that is better
            double curFitness = optimizee.evaluateFitness(p);
            int ct = 0;
            while (optimizee.evaluateFitness(nbr) > curFitness && ct < MAX_NBRS_TO_EXPLORE) {
                nbr = p.getRandomNeighbor(rad);
                ct++;
            }
        }
        return nbr;
    }

    /**
     * Evaluate the members of the population - either directly, or by
     * comparing them against the initial params value passed in (including params).
     * Note: this method assigns a fitness value to each member of the population.
     *
     * @param population the population to evaluate
     * @param previousBest the best solution from the previous iteration
     * @return the new best solution.
     */
    protected ParameterArray evaluatePopulation(List<ParameterArray> population, ParameterArray previousBest) {
        ParameterArray bestFitness = previousBest;

        for (ParameterArray p : population) {

            double fitness;
            if (optimizee.evaluateByComparison()) {
                fitness = optimizee.compareFitness(p, previousBest);
            } else {
                fitness = optimizee.evaluateFitness(p);
            }
            p.setFitness(fitness);
            if (fitness < bestFitness.getFitness()) {
                bestFitness = p;
                // show it if better than what we had before
                notifyOfChange(p);
                ThreadUtil.sleep(500);
            }
        }
        return bestFitness.copy();
    }

    private void printPopulation(List population) {
        printPopulation(population, population.size());
    }

    private void printPopulation(List population, int limit)  {
        for (int i=0; i<population.size() && i<limit; i++)
            System.out.println( i + ": " + population.get(i));
        System.out.println( "" );
    }
}

File: scala-source/com/barrybecker4/optimization/strategy/GeneticSearchStrategy.scala
// Copyright by Barry G. Becker, 2000-2026. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.strategy

import com.barrybecker4.optimization.optimizee.Optimizee
import com.barrybecker4.common.format.FormatUtil
import com.barrybecker4.math.MathUtil
import com.barrybecker4.optimization.parameter.{ParameterArray, ParameterArrayWithFitness, ParameterCrossover}
import ParameterArrayWithFitness.given
import scala.collection.mutable.ArrayBuffer
import scala.util.Random

object GeneticSearchStrategy {
  // Percent amount to decimate the parent population by on each iteration
  private val CULL_FACTOR = 0.7
  private val NBR_RADIUS = 0.1
  private val NBR_RADIUS_SHRINK_FACTOR = 0.8
  private val NBR_RADIUS_EXPAND_FACTOR = 1.1
  private val NBR_RADIUS_SOFTENER = 20.0
  private val INITIAL_RADIUS = 1.0
  private val MAX_NBRS_TO_EXPLORE = 8

  /** Probability of using crossover vs neighbor mutation when filling the population after culling. */
  private val CROSSOVER_PROBABILITY = 0.65

  /** this prevents us from running forever.  */
  private val MAX_ITERATIONS = 100
  /** If more than this many iterations with no improvement, then stop */
  private val MAX_ITER_NO_IMPRV = 3

  /** stop when the avg population score does not improve by better than this  */
  private[strategy] val DEFAULT_IMPROVEMENT_EPS = 0.000000000001
}

/**
  * Genetic Algorithm (evolutionary) optimization strategy.
  * Many different strategies are possible to alter the population for each successive iteration.
  * The 2 primary ones that I use here are unary mutation and cross-over.
  * See Chapter 6 in "How to Solve it: Modern Heuristics" for more info.
  *
  * Use a hardcoded static data interface to initialize so it can be easily run in an applet without using resources.
  * @param optimizee the thing to be optimized.
  * @param config    per-instance genetic options (defaults preserve prior behavior).
  * @author Barry Becker
  */
class GeneticSearchStrategy(
    optimizee: Optimizee,
    rnd: Random = MathUtil.RANDOM,
    val config: GeneticConfig = GeneticConfig()
) extends OptimizationStrategy(optimizee) {

  /** radius to look for neighbors in  */
  private var nbrRadius = GeneticSearchStrategy.NBR_RADIUS

  /** This is the desired number of members to be maintained in the population at any time.
    * Might not always get this many if there are duplicates or the search space is small.
    */
  private var desiredPopulationSize = 0

  /** If we don't improve by at least this amount between iterations, terminate.  */
  protected var improvementEpsilon: Double = GeneticSearchStrategy.DEFAULT_IMPROVEMENT_EPS

  def setImprovementEpsilon(eps: Double): Unit = {
    improvementEpsilon = eps
  }

  /** Finds a local minima using a genetic algorithm (evolutionary) search.
    * Stop iterating as soon as the average evaluation score of the population
    * does not significantly improve.
    * @param params the initial value for the parameters to optimize.
    * @param fitnessRange the approximate absolute value of the fitnessRange.
    * @return the optimized params.
    */
  override def doOptimization(params: ParameterArray, fitnessRange: Double): ParameterArrayWithFitness = {
    var lastBest: ParameterArrayWithFitness = null
    desiredPopulationSize = params.getSamplePopulationSize

    val population = findInitialPopulation(params)
    //println("The population of this generation is " + population.size()
    //    + " (desired was " + desiredPopulationSize + ")");
    // EVALUATE POPULATION
    lastBest = population.min //evaluatePopulation(population, params)
    evolve(lastBest, population, fitnessRange)
  }

  /** Visible for tests in `com.barrybecker4.optimization`. */
  private[optimization] def findInitialPopulation(params: ParameterArray): ArrayBuffer[ParameterArrayWithFitness] = {
    if (desiredPopulationSize <= 0) desiredPopulationSize = params.getSamplePopulationSize
    // create an initial population based on params and POPULATION_SIZE-1 other random candidate solutions.
    val population = ArrayBuffer[ParameterArrayWithFitness]()
    val fitness =
      if (optimizee.evaluateByComparison) Double.MaxValue
      else optimizee.evaluateFitness(params)
    val p1 = ParameterArrayWithFitness(params, fitness)
    population += p1
    var i = 0
    val max = 100 * desiredPopulationSize
    while (population.size < desiredPopulationSize && i < max) {
      val nbr = params.getRandomNeighbor(GeneticSearchStrategy.INITIAL_RADIUS)
      if (!population.contains(nbr)) {
        val fitness =
          if (optimizee.evaluateByComparison) optimizee.compareFitness(nbr, params)
          else optimizee.evaluateFitness(nbr)
        population += ParameterArrayWithFitness(nbr, fitness)
      }
      i += 1
    }
    if (population.size <= 1) throw new IllegalStateException("No random neighbors found for " + params)
    population
  }

  /** Runs one refill pass as in [[evolve]]; visible for tests. */
  private[optimization] def refillAfterCullForTesting(
      population: ArrayBuffer[ParameterArrayWithFitness],
      keepSize: Int,
      generationBaseline: ParameterArray,
      desiredPopSize: Int
  ): Unit = {
    val saved = desiredPopulationSize
    desiredPopulationSize = desiredPopSize
    try replaceCulledWithKeeperVariants(population, keepSize, generationBaseline)
    finally desiredPopulationSize = saved
  }

  /** Find the new best candidate.
    * @return the new best candidate.
    */
  private def evolve(lastBest: ParameterArrayWithFitness,
                          population: ArrayBuffer[ParameterArrayWithFitness],
                          fitnessRange: Double): ParameterArrayWithFitness = {
    var currentBest = lastBest
    var ct = 0
    var deltaFitness = -1.0
    var numWithNoImprovement = 0

    val lowThresh = 0.005 * fitnessRange
    val highThresh = 0.05 * fitnessRange

    var recentBest = lastBest
    var pop = population
    while (!shouldStopEvolution(deltaFitness, numWithNoImprovement, currentBest, ct)) {
      val generationBaseline = recentBest.pa
      reevaluatePopulationFitness(pop, generationBaseline)
      recentBest = syncBestAfterReevaluate(pop, generationBaseline)
      pop = cullPopulation(pop)
      replaceCulledWithKeeperVariants(pop, pop.size, generationBaseline)
      currentBest = pop.min
      trace("currBest = " + currentBest + " \nrecBest = " + recentBest + "        ct=" + ct)
      deltaFitness = computeFitnessDelta(recentBest, currentBest, ct)
      trace("delta fitness =" +
        deltaFitness + "      rbrRadius = " + nbrRadius + "  improvementEpsilon = " + improvementEpsilon)
      scaleNeighborRadius(deltaFitness, lowThresh, highThresh)
      recentBest = currentBest
      notifyOfChange(currentBest)
      ct += 1
      assert(deltaFitness <= 0, "The fitness should never get worse.")
      if (deltaFitness > -improvementEpsilon) {
        numWithNoImprovement += 1
      }
    }

    logEvolveTermination(currentBest, deltaFitness, numWithNoImprovement, ct)
    trace("----------------------- done -------------------")
    log(ct, currentBest, 0, 0, FormatUtil.formatNumber(ct))
    currentBest
  }

  /** After re-scoring vs `baselinePa`, refresh the tracked champion so its `fitness` matches the buffer. */
  private def syncBestAfterReevaluate(
      pop: ArrayBuffer[ParameterArrayWithFitness],
      baselinePa: ParameterArray
  ): ParameterArrayWithFitness =
    pop.find(_.pa eq baselinePa).getOrElse(pop.min)

  private def shouldStopEvolution(deltaFitness: Double, numWithNoImprovement: Int,
                                  currentBest: ParameterArrayWithFitness, ct: Int): Boolean =
    (deltaFitness >= -improvementEpsilon && numWithNoImprovement > GeneticSearchStrategy.MAX_ITER_NO_IMPRV) ||
      isOptimalFitnessReached(currentBest) ||
      ct > GeneticSearchStrategy.MAX_ITERATIONS

  private def scaleNeighborRadius(deltaFitness: Double, lowThresh: Double, highThresh: Double): Unit = {
    val factor =
      if (deltaFitness < -highThresh) GeneticSearchStrategy.NBR_RADIUS_EXPAND_FACTOR
      else if (deltaFitness > -lowThresh) GeneticSearchStrategy.NBR_RADIUS_SHRINK_FACTOR
      else 1.0
    nbrRadius *= factor
  }

  private def logEvolveTermination(currentBest: ParameterArrayWithFitness, deltaFitness: Double,
                                   numWithNoImprovement: Int, ct: Int): Unit = {
    if (isOptimalFitnessReached(currentBest))
      trace("stopped because we found the optimal fitness.")
    else if (deltaFitness >= -improvementEpsilon && numWithNoImprovement > GeneticSearchStrategy.MAX_ITER_NO_IMPRV) {
      trace("stopped because we made no IMPROVEMENT. The delta, " +
        deltaFitness + " was >= " + -improvementEpsilon)
    }
    else if (ct > GeneticSearchStrategy.MAX_ITERATIONS)
      trace(s"Stopped because we exceeded the MAX ITERATIONS(${GeneticSearchStrategy.MAX_ITERATIONS}): num iterations = $ct")
    else
      throw new IllegalStateException(s"stopped for unexpected cause. ct = $ct deltaFit=$deltaFitness " +
        s"currBest=$currentBest nbrRad=$nbrRadius")
  }

  /** Computes the fitness delta, but also logs and asserts that it is not 0.
    * @return the different in fitness between current best and last best.
    */
  private def computeFitnessDelta(lastBest: ParameterArrayWithFitness,
                                  currentBest: ParameterArrayWithFitness, ct: Int): Double = {
    val deltaFitness = currentBest.fitness - lastBest.fitness
    assert(deltaFitness <= 0, "We must never get worse in a new generation. Old fitness=" +
      lastBest.fitness + " New Fitness = " + currentBest.fitness + ".")
    log(ct, currentBest, nbrRadius, deltaFitness, "---")
    deltaFitness
  }

  /** Remove all but the best candidates. Better candidates have lower values.
    * @param population the whole population. It will be reduced in size.
    * @return the culled population
    */
  private def cullPopulation(population: ArrayBuffer[ParameterArrayWithFitness]) = {
    // sort the population according to the fitness of members.
    val sortedPopulation = population.sorted
    // throw out the bottom CULL_FACTOR*desiredPopulationSize members - keeping the cream of the crop.
    // then replace those culled with crossover and/or neighbor mutation of the parents that remain.
    val keepSize = Math.max(1, (sortedPopulation.size * (1.0 - GeneticSearchStrategy.CULL_FACTOR)).toInt)
    val size = sortedPopulation.size
    val culledPop = sortedPopulation.dropRight(size - keepSize)
    //println("new pop size = " + culledPop.size)
    //println("pop after culling: first = "  +  culledPop.get(0) + " last(" + culledPop.size() + ")");
    //printPopulation(sortedPopulation, 5);
    culledPop
  }

  /** Replace culled slots until `desiredPopulationSize` is reached (at most one new individual per attempt).
    *
    * @param generationBaseline compare-mode fitness for all new evaluations this generation uses this baseline
    */
  private def replaceCulledWithKeeperVariants(population: ArrayBuffer[ParameterArrayWithFitness],
                                              keepSize: Int,
                                              generationBaseline: ParameterArray): Unit = {
    val maxAttempts =
      math.max(100 * desiredPopulationSize, (desiredPopulationSize - keepSize) * 50 + 50)
    var attempts = 0
    while (population.size < desiredPopulationSize && attempts < maxAttempts) {
      attempts += 1
      val tryCross =
        config.crossoverEnabled && rnd.nextDouble() < GeneticSearchStrategy.CROSSOVER_PROBABILITY && keepSize >= 2
      var added = false
      if (tryCross) {
        var i1 = (rnd.nextDouble() * rnd.nextDouble() * keepSize).toInt
        var i2 = (rnd.nextDouble() * rnd.nextDouble() * keepSize).toInt
        if (keepSize > 1) {
          while (i2 == i1) i2 = (i2 + 1) % keepSize
        }
        val p1 = population(i1).pa
        val p2 = population(i2).pa
        if (p1.getClass == p2.getClass) {
          try {
            val child = ParameterCrossover.cross(p1, p2, rnd)
            if (!population.exists(_.pa == child)) {
              population += evaluateParameterArray(child, generationBaseline)
              notifyOfChange(population(i1))
              added = true
            }
          } catch {
            case _: IllegalArgumentException => ()
          }
        }
      }
      if (!added) {
        val r = rnd.nextDouble()
        val keeperIndex: Int = (r * r * keepSize).toInt
        val p = population(keeperIndex)
        val r2 = (keeperIndex + GeneticSearchStrategy.NBR_RADIUS_SOFTENER) / GeneticSearchStrategy.NBR_RADIUS_SOFTENER * nbrRadius
        val nbr = getNeighbor(p.pa, r2, generationBaseline)
        if (!population.exists(_.pa == nbr.pa)) {
          population += nbr
          notifyOfChange(p)
        }
      }
    }
  }

  private def evaluateParameterArray(pa: ParameterArray, generationBaseline: ParameterArray): ParameterArrayWithFitness = {
    val fitness =
      if (optimizee.evaluateByComparison) optimizee.compareFitness(pa, generationBaseline)
      else optimizee.evaluateFitness(pa)
    ParameterArrayWithFitness(pa, fitness)
  }

  /** @param p parameter array to get neighbor for
    * @param rad larger radius means more distant neighbor
    * @param generationBaseline compare-mode baseline for this generation (same as reevaluate/crossover)
    * @return a neighbor of p. If using absolute fitness, try to find a neighbor that has better fitnesss.
    */
  private def getNeighbor(p: ParameterArray, rad: Double, generationBaseline: ParameterArray): ParameterArrayWithFitness = {
    var nbr = p.getRandomNeighbor(rad)
    var nbrFitness = Double.MaxValue
    if (!optimizee.evaluateByComparison) { // try to find a nbr with fitness that is better
      val curFitness = optimizee.evaluateFitness(p)
      var ct = 0
      nbrFitness = optimizee.evaluateFitness(nbr)
      while (nbrFitness >= curFitness && ct < GeneticSearchStrategy.MAX_NBRS_TO_EXPLORE) {
        nbr = p.getRandomNeighbor(rad)
        nbrFitness = optimizee.evaluateFitness(nbr)
        ct += 1
      }
    } else {
      nbrFitness = optimizee.compareFitness(nbr, generationBaseline)
    }
    ParameterArrayWithFitness(nbr, nbrFitness)
  }

  /** Refresh per-individual fitness when using relative fitness vs the previous generation best. */
  protected def reevaluatePopulationFitness(population: ArrayBuffer[ParameterArrayWithFitness],
                                            baseline: ParameterArray): Unit = {
    if (!optimizee.evaluateByComparison) return
    var i = 0
    while (i < population.size) {
      val p = population(i).pa
      val f = optimizee.compareFitness(p, baseline)
      population(i) = ParameterArrayWithFitness(p, f)
      i += 1
    }
  }
}

/**
  * Per-instance options for [[GeneticSearchStrategy]] (thread-safe).
  *
  * @param crossoverEnabled when false, refill uses neighbor mutation only
  */
case class GeneticConfig(crossoverEnabled: Boolean = true)
